### PR TITLE
Dtspb 958 standard caveat e2 e

### PR DESCRIPTION
**Pull request by phaniperla**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/DTSPB-958

### Change description ###

Caveat E2E Tests

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```

### CODE REVIEW

**Review by arun-hmcts — APPROVED** [review]

**Review by Ferdi-Cocelli — APPROVED** [review]

File: test/end-to-end/helpers/JSWait.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                newUrl = helper.options.url + newUrl;
+            }
+
+            helper.page.goto(newUrl).catch(err => {
+                console.error(err.message);
+            });
+            await helper.page.waitForNavigation({waitUntil: 'networkidle0'});
+
+        } else {
+            await helper.amOnPage(newUrl);
```
Comment (vinceholt) [inline test/end-to-end/helpers/JSWait.js:44]: I used waitUrlEquals/waitInUrl in frontend code for this

File: test/end-to-end/pages/applicant/address.js [hunk 2]
```diff
@@ -1,17 +1,31 @@
...
-    I.fillField('#postTown', 'test address for applicant town');
-    I.fillField('#newPostCode', 'postcode');
-    I.waitForNavigationToComplete(`input[value="${commonContent.saveAndContinue}"]`);
-};
+    if (!testAddressIndex) {
+        testAddressIndex = '0';
+    }
+
+    I.waitInUrl(pageUnderTest.getUrl());
+    I.seeInCurrentUrl(pageUnderTest.getUrl());
```
Comment (vinceholt) [inline test/end-to-end/pages/applicant/address.js:18]: need await in all these I. methods

File: test/end-to-end/pages/applicant/address.js [hunk 3]
```diff
@@ -1,17 +1,31 @@
...
+    I.fillField('postcode', 'SW9 9PD');
+    await I.navByClick(findAddress);
+    I.wait(2);
+    I.waitForVisible('#postcodeAddress');
+    I.selectOption('#postcodeAddress', testAddressIndex);
+    I.wait(2);
+    I.waitForElement('#addressLine1');
```
Comment (vinceholt) [inline test/end-to-end/pages/applicant/address.js:25]: don't need I.wait(2), use e.g. await I.waitForElement(#addressLine1, config.TestTimeToWaitForElement), with a max time in TestTimeToWaitForElement

File: test/end-to-end/pages/applicant/email.js [hunk 4]
```diff
@@ -1,11 +1,16 @@
...
 const pageUnderTest = require('app/steps/ui/applicant/email/index');
 
-module.exports = function (email) {
+async function enterApplicantEmail(language, email) {
+    const commonContent = language === 'en' ? commonContentEn : commonContentCy;
     const I = this;
-    I.seeCurrentUrlEquals(pageUnderTest.getUrl());
+
+    I.seeInCurrentUrl(pageUnderTest.getUrl());
     I.fillField('#email', email);
```
Comment (vinceholt) [inline test/end-to-end/pages/applicant/email.js:12]: I would add awaits here, always think good practice with a browser component to wait for code to finish before continuing to next action

File: test/end-to-end/pages/applicant/name.js [hunk 5]
```diff
@@ -1,12 +1,17 @@
...
 const pageUnderTest = require('app/steps/ui/applicant/name/index');
 
-module.exports = function (firstname, lastname) {
+async function enterApplicantName(language ='en', firstname, lastname) {
+    const commonContent = language === 'en' ? commonContentEn : commonContentCy;
     const I = this;
-    I.seeCurrentUrlEquals(pageUnderTest.getUrl());
+
+    I.seeInCurrentUrl(pageUnderTest.getUrl());
     I.fillField('#firstName', firstname);
```
Comment (vinceholt) [inline test/end-to-end/pages/applicant/name.js:12]: await

File: test/end-to-end/pages/deceased/address.js [hunk 6]
```diff
@@ -1,17 +1,30 @@
...
-    I.fillField('#addressLine3', 'test address for deceased line 3');
-    I.fillField('#postTown', 'test address for deceased town');
-    I.fillField('#newPostCode', 'postcode');
-    I.waitForNavigationToComplete(`input[value="${commonContent.saveAndContinue}"]`);
-};
+    if (!testAddressIndex) {
+        testAddressIndex = '0';
+    }
+
+    I.waitInUrl(pageUnderTest.getUrl());
```
Comment (vinceholt) [inline test/end-to-end/pages/deceased/address.js:17]: await

File: test/end-to-end/pages/deceased/address.js [hunk 7]
```diff
@@ -1,17 +1,30 @@
...
+    I.seeInCurrentUrl(pageUnderTest.getUrl());
+    I.fillField('postcode', 'SW9 9PE');
+    await I.navByClick(findAddress);
+    I.wait(2);
+    I.waitForVisible('#postcodeAddress');
+    I.selectOption('#postcodeAddress', testAddressIndex);
+    I.wait(2);
+    I.waitForElement('#addressLine1');
```
Comment (vinceholt) [inline test/end-to-end/pages/deceased/address.js:25]: as above, no need for hardcoded waits, use explicit:
 await waitForElement

File: test/end-to-end/pages/deceased/dob.js [hunk 8]
```diff
@@ -1,14 +1,19 @@
...
 
-module.exports = function (day, month, year) {
+async function enterDeceasedDateOfBirth(language, day, month, year) {
+
+    const commonContent = language === 'en' ? commonContentEn : commonContentCy;
     const I = this;
 
-    I.seeCurrentUrlEquals(pageUnderTest.getUrl());
+    I.seeInCurrentUrl(pageUnderTest.getUrl());
     I.fillField('#dob-day', day);
```
Comment (vinceholt) [inline test/end-to-end/pages/deceased/dob.js:13]: awaits

File: test/end-to-end/pages/deceased/dobknown.js [hunk 9]
```diff
@@ -1,12 +1,16 @@
...
+async function enterDeceasedDateOfBirthKnown(language ='en') {
+    const commonContent = language === 'en' ? commonContentEn : commonContentCy;
     const I = this;
 
-    I.seeCurrentUrlEquals(pageUnderTest.getUrl());
-    I.click(`#dobknown-option${answer}`);
-    I.waitForNavigationToComplete(`input[value="${commonContent.saveAndContinue}"]`);
-};
+    I.seeInCurrentUrl(pageUnderTest.getUrl());
+    I.click('#dobknown');
```
Comment (vinceholt) [inline test/end-to-end/pages/deceased/dobknown.js:12]: awaits

File: test/end-to-end/pages/deceased/dod.js [hunk 10]
```diff
@@ -1,14 +1,19 @@
...
+async function enterDeceasedDateOfDeath(language ='en', day, month, year) {
+
+    const commonContent = language === 'en' ? commonContentEn : commonContentCy;
     const I = this;
 
-    I.seeCurrentUrlEquals(pageUnderTest.getUrl());
+    I.seeInCurrentUrl(pageUnderTest.getUrl());
     I.fillField('#dod-day', day);
     I.fillField('#dod-month', month);
     I.fillField('#dod-year', year);
```
Comment (vinceholt) [inline test/end-to-end/pages/deceased/dod.js:15]: awaits

File: test/end-to-end/pages/deceased/othernames.js [hunk 11]
```diff
@@ -1,25 +1,33 @@
...
     const I = this;
-    I.seeCurrentUrlEquals(pageUnderTest.getUrl());
+
+    I.seeInCurrentUrl(pageUnderTest.getUrl());
     let i = 1;
 
     while (i <= noOfAliases) {
         if (i === 1) {
             I.fillField('#otherNames_name_'+ (i-1) + '_firstName', 'alias_firstnames_' + i);
             I.fillField('#otherNames_name_'+ (i-1) + '_lastName', 'alias_lastnames_' + i);
```
Comment (vinceholt) [inline test/end-to-end/pages/deceased/othernames.js:20]: would recommend template literals `${}` as oppose to concat

File: test/end-to-end/pages/deceased/othernames.js [hunk 12]
```diff
@@ -1,25 +1,33 @@
...
+    const otherNames = language === 'en' ? otherNamesEn : otherNamesCy;
     const I = this;
-    I.seeCurrentUrlEquals(pageUnderTest.getUrl());
+
+    I.seeInCurrentUrl(pageUnderTest.getUrl());
     let i = 1;
 
     while (i <= noOfAliases) {
         if (i === 1) {
             I.fillField('#otherNames_name_'+ (i-1) + '_firstName', 'alias_firstnames_' + i);
```
Comment (vinceholt) [inline test/end-to-end/pages/deceased/othernames.js:19]: await

File: test/end-to-end/pages/deceased/othernames.js [hunk 13]
```diff
@@ -1,25 +1,33 @@
...
             I.fillField('#otherNames_name_'+ (i-1) + '_lastName', 'alias_lastnames_' + i);
         } else {
-            I.click('Add another name');
-            I.wait(10);
+            I.click(otherNames.addAnotherName);
+            I.wait(6);
             I.fillField('#otherNames_name_'+ (i-1) + '_firstName', 'alias_firstnames_' + i);
```
Comment (vinceholt) [inline test/end-to-end/pages/deceased/othernames.js:24]: remove hardcoded wait and use explicit: await I.click

File: test/end-to-end/pages/payment/govukconfirmpayment.js [hunk 14]
```diff
@@ -2,10 +2,11 @@
 
 const testConfig = require('config');
 
-module.exports = function () {
+async function seeGovUkConfirmPage () {
     const I = this;
 
     I.seeInCurrentUrl(testConfig.TestGovUkConfirmPaymentUrl);
+    await I.retry(2).navByClick('#confirm');
```
Comment (vinceholt) [inline test/end-to-end/pages/payment/govukconfirmpayment.js:9]: interested why we need a retry, probably need an await I.waitForElement prior to  click, I always add that in prior to clicking a button. you've also got waitForVisible and waitForClickable. more scientific and reliable 
than blunt retry

File: test/end-to-end/pages/payment/govukpayment.js [hunk 15]
```diff
@@ -15,5 +17,7 @@ module.exports = function () {
     I.fillField('#address-postcode', testConfig.govPayTestCardDetails.addressPostcode);
     I.fillField('#email', testConfig.TestEnvEmailAddress);
```
Comment (vinceholt) [inline test/end-to-end/pages/payment/govukpayment.js:19]: awaits

File: test/end-to-end/pages/startpage/eauality.js [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+
+    const I = this;
+    const stepContent = language === 'en' ? equalityEn : equalityCy;
+    I.wait(3);
+
+    const url = await I.grabCurrentUrl();
+
+    if (url.startsWith(pcqAAT)) {
+        I.waitInUrl(pagePath);
+        I.seeCurrentUrlEquals(pagePath);
```
Comment (vinceholt) [inline test/end-to-end/pages/startpage/eauality.js:16]: awaits

File: test/end-to-end/paths/caveat.e2e.js [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+        await I.seePaymentBreakdownPage(language);
+        if (testConfigurator.getUseGovPay() === 'true') {
+            await I.seeGovUkPaymentPage();
+            await I.seeGovUkConfirmPage();
+        }
+
+        const isCaseIDGenerated = await I.checkElementExist('//h1[contains(text(), \'Application complete\')]');
+        if (isCaseIDGenerated) {
+            I.seeThankYouPage();
+        }
```
Comment (vinceholt) [inline test/end-to-end/paths/caveat.e2e.js:43]: await

File: test/end-to-end/paths/caveat.e2e.js [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+
+    }).tag('@e2e')
+        .retry(2);
+
```
Comment (vinceholt) [inline test/end-to-end/paths/caveat.e2e.js:47]: num retries in config?

File: test/end-to-end/paths/caveat.e2e.js [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+        if (testConfigurator.getUseGovPay() === 'true') {
+            await I.seeGovUkPaymentPage();
+            await I.seeGovUkConfirmPage();
+        }
+
+        const isCaseIDGenerated = await I.checkElementExist('//h1[contains(text(), \'Application complete\')]');
+        if (isCaseIDGenerated) {
+            I.seeThankYouPage();
+        }
```
Comment (vinceholt) [inline test/end-to-end/paths/caveat.e2e.js:78]: await

**Review by vinceholt — COMMENTED** [review]
done a great job with introducing async / await pattern; just a bit more to go!

**Pull request merged**